rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,815 @@
+        sentence-level features, yielding two tensors (one for each feature type).
+    2. Concatenate the sequence- and sentence-level features into one tensor.
+
+    Arguments:
+        attribute: Name of attribute (e.g. `text` or `label`) whose features will be
+            processed.
+        attribute_signature: A dictionary containing two lists of `FeatureSignature`s, 
+            one for each feature type of the given attribute.
+        config: A model config used for correctly parametrising the the layer and the
+            `ConcatenateSparseDenseFeatures` layer it creates internally.
+
+    Input shape:
+        Tuple of four tensor inputs:
+            sequence_features: List of 3-D dense or sparse tensors, each with shape
+                `(batch_size, max_seq_length, input_dim)` where `input_dim` can be
+                different for sparse vs dense tensors.
+            sentence_features: List of 3-D dense or sparse tensors, each with shape
+                `(batch_size, 1, input_dim)` where `input_dim` can be different for
+                sparse vs dense tensors, and can differ from that in `sequence_features`.
+            sequence_feature_lengths: Dense tensor of shape `(batch_size, )` containing
+                the real sequence length for each example in the batch, i.e. the lengths
+                of the real (not padded) sequence-level (token-level) features.
+
+    Output shape:
+        combined features: a 3-D tensor with shape `(batch_size, sequence_length, units)`
+            where `units` is  completely  determined by the internally applied 
+            `ConcatenateSparseDenseFeatures` layer and `sequence_length` is the combined
+            length of sequence- and sentence-level features (`max_seq_length + 1` if both
+            sequence- and sentence-level features are present, `max_seq_length` if only 
+            sequence-level features are present, and 1 if only sentence-level features
+            are present).
+        mask_combined_sequence_sentence: a 3-D tensor with shape 
+            `(batch_size, sequence_length, 1)`.
+        
+    Raises:
+        A ValueError if no feature signatures are provided.
+    """
+
+    def __init__(
+        self,
+        attribute: Text,
+        attribute_signature: Dict[Text, List[FeatureSignature]],
+        config: Dict[Text, Any],
+    ) -> None:
+        if not attribute_signature or not (
+            attribute_signature.get(SENTENCE, [])
+            or attribute_signature.get(SEQUENCE, [])
+        ):
+            raise ValueError(
+                "The attribute signature must contain some feature signatures."
+            )
+
+        super().__init__(name=f"rasa_feature_combining_layer_{attribute}")
+
+        self._tf_layers = {}
+
+        # Prepare layers for combining sparse and dense features for each feature type
+        self._feature_types_present = self._prepare_sparse_dense_concat_layers(
```

Comment:
I think the name of the method doesn't correspond to the name of the output

Reply:
Yep. And this is not the only place like this. Basically, I was struggling to come up with a sensible flow that wouldn't duplicate the logic. The problem is: I need to prepare some layers. In the process of preparing them, I also necessarily have to construct some variable (in this case the `feature_types_present` dict) which is useful also outside of the preparation method. Now that I'm giving it a second thought, I think I'll write it out in a more verbose way. Something like:
```python
self._feature_types_present = self._get_present_feature_types()
self._prepare_sparse_dense_concat_layers() # this will internally use self._feature_types_present
```

Reply:
@wochinge no need to check the whole PR, but could you please help us out here. What do you think would be the best approach?

Reply:
@Ghostvv I'm curious what you think about `_prepare_sequence_sentence_concat`. Because it outputs `output_units` as a by-product, which isn't reflected in the method's name. But the logic for creating sequence-sentence combining components and for determining `output_units` is so intertwined that I think I'd literally have do duplicate most of it if I were to have two methods (`_prepare_sequence_sentence_concat` and `_calculate_output_units`). At the same time, I think `_prepare_sequence_sentence_concat` describes the most important thing that the method does...

Reply:
yeah, I don't know. I see that we need to couple these calculations. I think if we come up with alternative verb to `prepare` it'll be fine

Reply:
Sorry, didn't have time before my vacation to check it. Can you please reach out to somebody else from Enable? 🙌🏻

Reply:
@alwx could you help us? It's just about this one review comment.

Reply:
I think the struggle might come from the fact that the method is doing two things. It finds the present feature types and returns them, but as a side effect it also initializes the layers. Might be easier to understand and to name if you split it up: something like `find/collect/determine_feature_types` where you just get the feature types dict. And then a second method `initialize/instantiate_concatenation_layers` which uses this dict or so.

What do you think?

Reply:
Oh, no, sorry, @twerkmeister that issue is now resolved (in line with what you suggested). I meant my question in the [specific comment](https://github.com/RasaHQ/rasa/pull/7589#discussion_r574446359).